In [3]:
import mpp_classifiers as mpp
import load_ship_data as lsd
import numpy as np
import time
from sklearn.decomposition import PCA
from sklearn.model_selection import KFold

In [4]:

data_path="data/shipsnet.json"
train,test,valid=lsd.load_data_train_test_split(data_path)
print('''
Due to randomization in the data splitting process, accuracy and runtime of classifiers
will vary slightly between runs during our initial exploration and dimension reduction.
Once we find a promising dimension reduction and classifier combination, we integrate
K-Fold Cross Validation.''')

Training Set Data Length:  2800   Label Length:  2800
TestingSet Set Data Length:  600  Label Length:  600
Validation Set Data Length:  600  Label Length:  600

Due to randomization in the data splitting process, accuracy and runtime of classifiers
will vary slightly between runs during our initial exploration and dimension reduction.
Once we find a promising dimension reduction and classifier combination, we integrate
K-Fold Cross Validation.


In [5]:
ship_prob=np.sum(train[1])*1.0/train[1].shape[0]
p=np.array([1-ship_prob,ship_prob])
print(p)
Xtrain=train[0]
ytrain=train[1]
Xtest=test[0]
ytest=test[1]
Xvalid=valid[0]
yvalid=valid[1]
Xtrain=Xtrain.reshape(Xtrain.shape[0],Xtrain.shape[1]*Xtrain.shape[2]*Xtrain.shape[3])
Xtest=Xtest.reshape(Xtest.shape[0],Xtest.shape[1]*Xtest.shape[2]*Xtest.shape[3])
Xvalid=Xvalid.reshape(Xvalid.shape[0],Xvalid.shape[1]*Xvalid.shape[2]*Xvalid.shape[3])
print(Xtrain.shape)
print(Xtest.shape)
print(Xvalid.shape)

[0.74464286 0.25535714]
(2800, 19200)
(600, 19200)
(600, 19200)


In [4]:
t0 = time.time()           # start time
y_model = mpp.mpp(Xtrain, ytrain, Xtest, 1, p)
t1 = time.time()           # ending time
acc_classwise, acc_overall = mpp.accuracy_score(ytest, y_model)
print(f'Overall accuracy = {acc_overall};')
print(f'Classwise accuracy = {acc_classwise};')
print(f'The learning process takes {t1 - t0} seconds.')

Overall accuracy = 0.7566666666666667;
Classwise accuracy = [1. 0.];
The learning process takes 146.8991458415985 seconds.


In [5]:
mpp_1_0_p=p
mpp_1_0_acc=acc_overall
pmm_1_0_class=acc_classwise
mpp_1_0_time=t1-t0
p=np.array([.5,.5])
t0 = time.time()           # start time
y_model = mpp.mpp(Xtrain, ytrain, Xtest, 1, p)
t1 = time.time()           # ending time
acc_classwise, acc_overall = mpp.accuracy_score(ytest, y_model)
print(f'Overall accuracy = {acc_overall};')
print(f'Classwise accuracy = {acc_classwise};')
print(f'The learning process takes {t1 - t0} seconds.')
mpp_1_1_p=p
mpp_1_1_acc=acc_overall
pmm_1_1_class=acc_classwise
mpp_1_1_time=t1-t0

Overall accuracy = 0.685;
Classwise accuracy = [0.6938326  0.65753425];
The learning process takes 42.332788705825806 seconds.


In [6]:
p=np.array([.9,.1])
t0 = time.time()           # start time
y_model = mpp.mpp(Xtrain, ytrain, Xtest, 1, p)
t1 = time.time()           # ending time
acc_classwise, acc_overall = mpp.accuracy_score(ytest, y_model)
print(f'Overall accuracy = {acc_overall};')
print(f'Classwise accuracy = {acc_classwise};')
print(f'The learning process takes {t1 - t0} seconds.')
mpp_1_2_p=p
mpp_1_2_acc=acc_overall
pmm_1_2_class=acc_classwise
mpp_1_2_time=t1-t0

Overall accuracy = 0.7566666666666667;
Classwise accuracy = [1. 0.];
The learning process takes 41.83013391494751 seconds.


In [4]:
pca = PCA(n_components=1920)
pca.fit(Xtrain)

PCA(n_components=1920)

In [6]:
pca_Xtrain=pca.transform(Xtrain)

In [7]:
pca_Xtest=pca.transform(Xtest)

In [16]:
explained=np.sum(pca.explained_variance_ratio_)
print(explained)

0.9989776178004813


In [8]:
p=np.array([1-ship_prob,ship_prob])
t0 = time.time()           # start time
y_model = mpp.mpp(pca_Xtrain, ytrain, pca_Xtest, 1, p)
t1 = time.time()           # ending time
acc_classwise, acc_overall = mpp.accuracy_score(ytest, y_model)
print(f'Overall accuracy = {acc_overall};')
print(f'Classwise accuracy = {acc_classwise};')
print(f'The learning process takes {t1 - t0} seconds.')

Overall accuracy = 0.7583333333333333;
Classwise accuracy = [1. 0.];
The learning process takes 0.32712292671203613 seconds.


In [4]:
pca2 = PCA(n_components=192)
pca2.fit(Xtrain)

PCA(n_components=192)

In [5]:
pca2_Xtrain=pca2.transform(Xtrain)
pca2_Xtest=pca2.transform(Xtest)
explained=np.sum(pca2.explained_variance_ratio_)
print(explained)

0.941326471367284


In [6]:
p=np.array([1-ship_prob,ship_prob])
t0 = time.time()           # start time
y_model = mpp.mpp(pca2_Xtrain, ytrain, pca2_Xtest, 1, p)
t1 = time.time()           # ending time
acc_classwise, acc_overall = mpp.accuracy_score(ytest, y_model)
print(f'Overall accuracy = {acc_overall};')
print(f'Classwise accuracy = {acc_classwise};')
print(f'The learning process takes {t1 - t0} seconds.')

Overall accuracy = 0.7766666666666666;
Classwise accuracy = [1. 0.];
The learning process takes 0.03291034698486328 seconds.


In [7]:
p=np.array([1-ship_prob,ship_prob])
t0 = time.time()           # start time
y_model = mpp.mpp(pca2_Xtrain, ytrain, pca2_Xtest, 2, p)
t1 = time.time()           # ending time
acc_classwise, acc_overall = mpp.accuracy_score(ytest, y_model)
print(f'Overall accuracy = {acc_overall};')
print(f'Classwise accuracy = {acc_classwise};')
print(f'The learning process takes {t1 - t0} seconds.')

Overall accuracy = 0.9366666666666666;
Classwise accuracy = [0.97424893 0.80597015];
The learning process takes 0.8736593723297119 seconds.


In [8]:
p=np.array([1-ship_prob,ship_prob])
t0 = time.time()           # start time
y_model = mpp.mpp(pca2_Xtrain, ytrain, pca2_Xtest, 3, p)
t1 = time.time()           # ending time
acc_classwise, acc_overall = mpp.accuracy_score(ytest, y_model)
print(f'Overall accuracy = {acc_overall};')
print(f'Classwise accuracy = {acc_classwise};')
print(f'The learning process takes {t1 - t0} seconds.')

Overall accuracy = 0.79;
Classwise accuracy = [0.78755365 0.79850746];
The learning process takes 1.4650781154632568 seconds.


Observations for Report:

Even with a linear classifier, our flattened training data set of dimension [2800,19200] takes an excessive amount of time to classify. Euclidean distance (case 1) yielded an accuracy of 75.6667%. Classwise accuracy showed that case 1 was merely classifying all samples as class 0 (no ship). The training time on this, our fastest classifier, was 146.8991 seconds. Ergo, PCA and FLD will be critical to classifying this data in a timely fashion. Using PCA with 1920 dimensions (99.898% of variance explained) reduced the training time for case 1 from 146.8991 seconds to .3271 seconds without any loss of accuracy on case 1. However, this dimension reduction was insufficient for case 2, which still took an excessive amount of time to train. So we ran PCA again with dimension 192. This preserved 94.1326% of the variation, and yielded extremely quick classification by all three cases. Mahalanobis distance (case 2) performed the best by a large margin with 93.6667% accuracy. Classwise performace using Mahalanobis distance was also vastly improved from Euclidean distance with 97.4249% accuracy on class 0 (no ship) and 80.5970% accuracy on class 1 (ship). Quadratic distance performed only slightly better than Euclidean distance ( 79% accuracy vs 77.6667% accuracy) overall, but did much better in term of classwise performance ([.7876,.7985] vs [1,0]). 

Major Results: 

PCA+Mahalanobis yielded \~94% accuracy with excellent classwise accuracy (\~97%,\~81%) and runtime (\~.87 seconds) using 192 dimensions.

Dimension reduction is critical on dataset of this size.

#### Postprocessing
Now that we have a promising dimension reduction and classifier,

we apply K-Fold Cross-Validation to verify our results


In [7]:
pca2 = PCA(n_components=192)
Xtrain=np.concatenate((Xtrain,Xvalid))
ytrain=np.concatenate((ytrain,yvalid))
pca2.fit(Xtrain)
pca2_Xtrain=pca2.transform(Xtrain)
pca2_Xtest=pca2.transform(Xtest)
explained=np.sum(pca2.explained_variance_ratio_)
print(explained)

0.9379117138032105


In [10]:
p=np.array([1-ship_prob,ship_prob])
X=pca2_Xtrain
y=ytrain
print(X.shape)
print(y.shape)
k=10
kf=KFold(n_splits=k,random_state=None)
acc_scores=[]
for train_index,test_index in kf.split(X):
    X_train,X_test=X[train_index,:],X[test_index,:]
    y_train,y_test=y[train_index],y[test_index]
    y_model = mpp.mpp(X_train, y_train, X_test, 2, p)
    acc_classwise, acc_overall = mpp.accuracy_score(y_test, y_model)
    acc_scores.append(acc_overall)
avg_acc=sum(acc_scores)*1.0/k
print('Accuracy from each fold =  {}'.format(acc_scores))
print('Average Accuracy = {}'.format(avg_acc))

(3400, 192)
(3400,)
Accuracy from each fold =  [0.9058823529411765, 0.9058823529411765, 0.95, 0.9294117647058824, 0.95, 0.9382352941176471, 0.9117647058823529, 0.9147058823529411, 0.9117647058823529, 0.9235294117647059]
Average Accuracy = 0.9241176470588236


In [11]:
y_model = mpp.mpp(X_train, y_train, X_test, 2, p)
acc_classwise, acc_overall = mpp.accuracy_score(y_test, y_model)
print(acc_overall)

0.9235294117647059


Ergo, K-Fold Cross-Validation indicates that we can reliably expect 92% accuracy on our dataset

In [3]:
def conf(pred,y):
    T0=sum([1 if x==y and y==0 else 0 for (x,y) in zip(pred,y)])
    T1=sum([1 if x==y and y==1 else 0 for (x,y) in zip(pred,y)])
    F0=sum([1 if x!=y and y==0 else 0 for (x,y) in zip(pred,y)])
    F1=sum([1 if x!=y and y==1 else 0 for (x,y) in zip(pred,y)])
    return np.array([[T0,T1],[F0,F1]])

    